### ISSUE TITLE
A2G Tasking: New mission type idea: Aerial Bombardment

### ISSUE THREAD

**Issue by Mechanist11** [FlightControl-Master/MOOSE#650]
New mission idea for low-tech airplanes and strategic targets. Aerial Bombardment

Please see enclosed presentation for further details

[A2G_TASK_AB.pdf](https://github.com/FlightControl-Master/MOOSE/files/1207442/A2G_TASK_AB.pdf)

**Comment by FlightControl-Master** [comment]
Thanks @Mechanist11 great idea! But first i wanna get into cargo.

**Comment by Mechanist11** [comment]
Take you time, noone is in a hurry ;)

**Comment by FlightControl-Master** [comment]
time we have now!

**Comment by thebgpikester** [comment]
OK I reviewed the PDF and here are my comments. Note that my comments are based from knowing the tasking system, knowing the current implementations in MOOSE, the DCS World environment and limitations and having executed similar end goals successfully myself to date. So I'm basing this on experience not just randomly commenting.

Firstly, Polygon zones versus point. DCS always drops bombs on a point, via Bombing or Attack Map Object.
![image](https://user-images.githubusercontent.com/22999891/48345896-40f08e80-e671-11e8-8752-6e758f77943f.png)
The area around it may also be relevant, but note that it is now easy to have multiple scenery inside a very very small area. You can do a polyzone, but it should be very very small (see later on)

Now, there are some important constraints completely missed.

**You cannot enumerate Scenery objects into the MOOSE DB**. I need to explain why this is key. Since 2.0 the amount of scenery object in a DCS map is well into multiple millions. A single warehouse 3D object will be surrounded by objects such as parked cars, fire extinguishers on walls, rubbish bins, lamp posts, fencing and all sorts are all individual objects with a unique scenery number. An average 250bomb will generate a hit event for 4-5 objects. Less on Caucasus, more on PG and NTTR. Depending where you put the explosion.

The current get Scenery functions are not working in MOOSE. They never have worked. They were looking to enumerate Scenery ID's in a radius. I think this is still possible, but you need to keep it super small for reasons that will become apparent.

We can check for dead objects in an event handler in MOOSE with a simple event handler. The EVENT.Dead returns the scenery ID, and this is the info you get:
2018-11-12 12:16:10.765 INFO    SCRIPTING:   5385(    13)/E:               EVENT00000.onEvent({[1]="S_EVENT_DEAD",[2]={["IniUnit"]={["SceneryObject"]={["id_"]=989069663,},["ClassName"]="SCENERY",["SceneryName"]=989069663,},["initiator"]=,["id"]=8,["IniObjectCategory"]=5,["IniUnitName"]=989069663,["IniCategory"]=4,["time"]=25227.901,["IniTypeName"]="SCENERY",["IniDCSUnit"]=,["IniDCSUnitName"]=989069663,},[3]=989069663,[5]=-1,})

I believe you can get the coordinate from it, but it will not match the coordinate of your zone or point, for sure. You can probably pin it to an area that it occupies with other scenery of course.

So at best, you know something dies NEAR your area. But what?

So far, killing a bridge is not a simple task. 
Firstly every bridge is different. Different size.
Secondly, they are often made of separate pieces
Here's a mapping I did for the Colorado bridge in NTTR:

> --Bridge
> -- WEST end    78577701
> -- WEST side   78577721
> -- West inside 78577737
> -- Centre      78577761
> -- EAST inside 78577781
> -- EAST side   78577805
> -- EAST end    78577823

I had to place a zone and create an explosion small enough to hit the right object with an event handler, and write down each segment.
Once I had all the segments, I wrote an event handler to check if any of these was dead and then set a flag. (side note, event handlers with multiple IF's running in frame are expensive CPU, I really do not like using them in big missions, especially with sensitive to server fps VR guys)

So, the problem should be easy to see now. You simply don't know what got hit, before it get's hit and only if you designed it with the work you put in. Mapping the 3D model to the scenery id is a bit of a manual task. Certainly you aren't going to want to do a getVec3 on every unit id, all those millions. Maybe you would, maybe you could have a mapped table of locations, but what about names?

I "used to" get model names with the event. Here's a mapping I did for Laughlin key scenery ID's (note the runway is included)
--Airfield from South to North (doubles)
--135136919 - LAUGHLIN_AFB_TOWER (control tower south)
--98435277  - AIRBASEHANGAR01SMALL (H01)
--98435072  - Hendersonaiporthungar (NE A02)
--78578289  - HUNGAR_028 (Largest, N of A02) 
--78578249  - HUNGAR_014 (

--5000026 - Main Laughlin Runway

Those were model names from ED. There are lots of them. Problem is, they can also be individuals, see the LAUGHLIN_AFB_TOWER is clearly a special model written just for that airbase. So no clear patterns to follow. Teh hangars are likely generic to the scenery.

So I believe the way to solve this has to be simple and here is my thoughts on design.

In the Mission, the MD must place polyzones around the exact item he wants destroyed, in as small a "ring" as he can manage. The PolyZone should have a name that will be scanned at launch, when instantiating the poly zone.
![image](https://user-images.githubusercontent.com/22999891/48348014-edce0a00-e677-11e8-81de-7681f3434fed.png)
This zone on NTTR has over 270 individual scenery ID's in it. I tackled this type of "generic destruction" as a percentage based destruction. Check the relevant MIST code:
```
local polyzone = mist.getGroupPoints('polyzone')
local deadones = mist.getDeadMapObjsInPolygonZone(polyzone)
local num = table.getn(deadones)
--env.info("Count of dead objects in zone is: " .. num)
local percentdestruction = round(num/270*100,0)
if percentdestruction <75 then
MESSAGE:New("Bullhead Fort HQ is  " .. percentdestruction .. "% destroyed .",10):ToAll()
end
```
When the destruction amount reached a certain percent, i marked it as a success. It's neither exact or great, but for loose bombs, it does work, as long as you know roughly how many objects should be in the zone to begin with.

If you make a very small zone, around say a building, sure you can check the scenery is inside that, but honestly, the fire extinguishers break it a bit and it's not easy. I would still recommend the MD tests with an explosion in Zone, perhaps using the standard DCS circle ZONE, to check on numbers. Or we fix the scenery enumeration tests in MOOSE and run ascript that WARNS of multiple scenery in zone.

From here on, it's almost plain sailing, because the TASK system is already a module well built. It will rely on something other than detection and have to cope with managing "ticking zones off a list". If it considers the scenery to be destroyed in the zone, it can tell the FSM the task is complete.

**Comment by Mechanist11** [comment]
I think you mist the point entirely. There are no intention to use scenery objects at all. It is a simplified logic, you make a zone poly or radial and count hit events in them, no matter what is hit. So you need to deliver 6000 pounds of bombs in a zone, If you drop 3 Mk84-s in the same point the task would be complete.

**Comment by thebgpikester** [comment]
Yes I missed that entirely! 

So I've done this recently, adapted from when i read through Range Script and it's quite nasty on burst CPU due to real time event tracking. Let me explain my finding with code examples.

First thing you need to know is that Range script doesn't track the bomb hitting anything at all! What Range script does, is track the Weapon by coordinates until they cease to exist, many times a second and extrapolates the last known coordinate and uses that as the landing point. It begins with the on shot event. Ripples are painful, a full squadron of B52's or Flying Fortress, I don't want to know!

My attempt ignored the tracking because I started off using 50 parachutists I wanted to track. 50 weapons tracks is quite nasty on CPU. For things in this range I'd suggest reducing the timer cycle by a lot and trying to make it as asynchronous as possible.

All i checked was firstly, what type of weapon, and if it was a certain type (parachute box) then I checked the coordinate of the dropping plane was in a zone:

```

AnyShot = EVENTHANDLER:New()
AnyShot:HandleEvent( EVENTS.Shot )
function AnyShot:OnEventShot( EventData )
local ThisUnit = UNIT:FindByName(EventData.IniDCSUnitName)
local ThisGroup = GROUP:FindByName(EventData.IniDCSGroupName)
local Coalition = ThisGroup:GetCoalition()

if EventData.WeaponName == "weapons.bombs.AIRBORNE_3_PARASHUT_BOX" then 

if ThisUnit:IsInZone(AbuMusa) then
  
  if Coalition == 1 then
  RedCargoDrop=RedCargoDrop+1
    if RedCargoDrop==4 then
      MESSAGE:New("RED cargo drop successful in Abu Musa.", 15):ToAll()
      RedCargoDrop=0
      SpawnRedCargoAbuMusa()
    end
  elseif Coalition == 2 then ...
```
I have four zones and 2 coalitions doing this with one type of weapon and it still gives a little 100ms stutter in multiplayer, basically due to all the realtime nested "if's" I would have tracked the weapons landing, but this is when I discovered there is no hit event when a weapon doesn't hit anything.

I originally tried to use Paratroopers and ripples of 50 and the stutter was quite measurable. Range script really is designed to track one at a time anyway, rather than multiples. I suppose it could be slackened off but at the very least you have to check if the weapon began life close to your zones of interest, which means a check on the Shot event and a check on the weapon type and get the location, at least begining very similar to the script above.

If you want to adapt Range script _I'd_ dumb it down to my example script level and track the release point only. A mapping of Weapon type to kg Mass would then have to be available to add the "Mass in zone". That's a lot of work, because all the weapons data is in the DCS core lua and you cannot use this without making some static mapping of bomb to mass which is a lot of effort, the amount of weapons in game is a massive list. But without it, the Event.shot would have no idea what was dropped, just where.

I'm not a great fan of any approaches to this that I have seen or made. They almost always need to begin with EVENT's and then checks on the event to filter it which leads to costly CPU in frame. I think it's achievable on an easy scale but not scalable. It's also not measuring things accurately, like if the bombs were armed or actually HIT anything at all. Which leads to a Player experience that begins to not meet expectations. Recording a Hit in Zone and leaving the "Mass" calculations might be a softer approach.

**Comment by thebgpikester** [comment]
Thank you for making the effort to improve MOOSE. Our main developer Flight Control has not been contributing to MOOSE for over a year now and we are in sustained maintenance mode for emergency bug fixes for modules that he created. 
In order for us to be working on the most serious of issues we have to make some hard decisions on how to prioritise Github tickets. 
•	Currently, emergency issues created by DCS changes are our number one concern
•	Issues with wide and serious impact effect are our next concern but after so many years most have been found.
•	Any Modules created by FunkyFranky (Ops) will require less analysis and investigation and he will prioritise separately.
•	CORE MOOSE issues, will always be important but will often require long investigations on unfamiliar code.
•	Fixes on FlightControl’s FUNCTIONAL and TASK, DISPATCHER modules are incredibly complex to rework his original code.
•	Finally feature requests will in most cases have to be closed but will be assessed in case they are simple to do.
Github uses a system of labels and states for tickets that is not limited to closed/open and all your data uploaded, and details will remain viewable by everyone. Your work has not been lost or disregarded, but in this case, it’s fairer to set realistic expectations. As things change over time this will be reassessed.
